### PR TITLE
Using multiple backends and single service

### LINKED ISSUE #58: Cache manager shutdown before flushing offset
Currently, we stop the cache manager in a few ways, including by listening to the `ShutdownEvent`. When the application is shutting down and the Debezium engine stops, it will try to flush the current offset. It is possible to stop the `RemoteCacheManager` _before_ flushing the offset to the remote cache. This will raise an exception during shutdown:

```text
2022-12-28 16:49:36,320 INFO  [io.gin.cdc.rem.RemoteOffsetStore] (engine) Setting {java.nio.HeapByteBuffer[pos=0 lim=29 cap=29]=java.nio.HeapByteBuffer[pos=0 lim=54 cap=54]}
2022-12-28 16:49:36,321 INFO  [org.inf.HOTROD] (engine) ISPN004002: Cannot perform operations on a cache associated with an unstarted RemoteCacheManager. Use RemoteCacheManager.start before using the remote cache.
2022-12-28 16:49:36,323 INFO  [io.gin.cdc.rem.RemoteOffsetStore] (engine) Stopping remote offset store.
2022-12-28 16:49:36,325 INFO  [io.deb.emb.EmbeddedEngine] (main) Stopping the embedded engine
2022-12-28 16:49:36,325 INFO  [io.deb.emb.EmbeddedEngine] (main) Waiting for PT5M for connector to stop
2022-12-28 16:49:36,325 ERROR [io.gin.hea.ConnectorReadiness] (engine) Engine rule1 failed [Error Occurred After Shutdown]: org.infinispan.client.hotrod.exceptions.RemoteCacheManagerNotStartedException:: Cannot perform operations on a cache associated with an unstarted RemoteCacheManager. Use RemoteCacheManager.start before using the remote cache.
	at org.infinispan.client.hotrod.impl.RemoteCacheImpl.assertRemoteCacheManagerIsStarted(RemoteCacheImpl.java:587)
	at org.infinispan.client.hotrod.impl.RemoteCacheImpl.putAllAsync(RemoteCacheImpl.java:255)
	at org.infinispan.client.hotrod.impl.RemoteCacheSupport.putAllAsync(RemoteCacheSupport.java:53)
	at io.gingersnapproject.cdc.cache.hotrod.HotRodOffsetBackend.set(HotRodOffsetBackend.java:38)
	at io.gingersnapproject.cdc.remote.RemoteOffsetStore.set(RemoteOffsetStore.java:51)
	at org.apache.kafka.connect.storage.OffsetStorageWriter.doFlush(OffsetStorageWriter.java:175)
	at io.debezium.embedded.EmbeddedEngine.commitOffsets(EmbeddedEngine.java:1059)
	at io.debezium.embedded.EmbeddedEngine.run(EmbeddedEngine.java:910)
	at io.debezium.embedded.ConvertingEngineBuilder$2.run(ConvertingEngineBuilder.java:195)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

We should handle this scenario and only stop the cache manager after the Debezium engine stopped and the offset was flushed. We can use the `RemoteOffsetStore#stop` and issue an event while the cache manager waits for an event from all created caches before stopping.

### PR DESCRIPTION
**Pull request by jabolina**
Fixes #58

### CODE REVIEW

File: src/main/java/io/gingersnapproject/cdc/cache/hotrod/HotRodService.java [hunk 1]
```diff
@@ -38,70 +46,83 @@ public class HotRodService implements CacheService {
...
-   public CompletionStage<Boolean> reconnect() {
-      if (rcm == null) {
-         throw new IllegalStateException("RemoteCacheManager not initialized");
-      }
-      return rcm.startAsync().handle((__, t) -> {
+   public CompletionStage<Boolean> reconnect(String name, Rule rule) {
+      if (!backends.containsKey(name)) return CompletableFutures.completedFalse();
+
+      return CompletableFuture.supplyAsync(() -> {
+         var backend = backendForRule(name, rule);
```
Comment (ryanemerson) [inline src/main/java/io/gingersnapproject/cdc/cache/hotrod/HotRodService.java:70]: Redundant var?

File: src/main/java/io/gingersnapproject/cdc/cache/hotrod/HotRodService.java [hunk 2]
```diff
@@ -38,70 +46,83 @@ public class HotRodService implements CacheService {
...
+         backend.start();
+         return null;
+      }).handle((ignore, t) -> {
          boolean successful = t == null;
          if (successful) {
             createCaches();
-            eventing.backendStartedEvent(true);
+         } else {
+            eventing.backendFailedEvent(name, t);
```
Comment (ryanemerson) [inline src/main/java/io/gingersnapproject/cdc/cache/hotrod/HotRodService.java:78]: Nitpick, but IMO this is more readable as:

```java
         if (t == null) {
            createCaches();
            return true;
         } else {
            eventing.backendFailedEvent(name, t);
            return false;
         }
```

**Review by ryanemerson — COMMENTED** [review]
LGTM, just a couple of minor comments.

**Comment by jabolina** [comment]
Thannks for the review @ryanemerson. Applied all suggestions.

**Comment by ryanemerson** [comment]
Thanks @jabolina

**Pull request merged**